ntoskrnl/se/token.c

```diff
@@ -135,34 +135,138 @@ SepDeleteTokenLock(
+            {
+                DPRINT1("SepCompareTokens(): SIDs are not equal!\n");
+                goto Quit;
+            }
+
+            /* Are the attributes from either SID from tokens equal? */
+            if (FirstToken->UserAndGroups[FirstTokenIndex].Attributes !=
+                SecondToken->UserAndGroups[SecondTokenIndex].Attributes)
+            {
+                DPRINT1("SepCompareTokens(): SID attributes are not equal!\n");
+                goto Quit;
+            }
+        }
+    }
 
     Restricted = SeTokenIsRestricted(FirstToken);
     if (Restricted == SeTokenIsRestricted(SecondToken))
     {
         if (Restricted)
         {
-            /* FIXME: Check if every SID that is restricted in either token is also restricted in the other one */
+            /* Check if every SID that is restricted in either token is also restricted in the other one */
+            for (FirstTokenIndex = 0; FirstTokenIndex < FirstToken->RestrictedSidCount; FirstTokenIndex++)
+            {
+                for (SecondTokenIndex = 0; SecondTokenIndex < SecondToken->RestrictedSidCount; SecondTokenIndex++)
+                {
+                    if (!RtlEqualSid(FirstToken->RestrictedSids[FirstTokenIndex].Sid,
+                                     SecondToken->RestrictedSids[SecondTokenIndex].Sid))
+                    {
+                        DPRINT1("SepCompareTokens(): Restricted SIDs are not equal!\n");
+                        goto Quit;
+                    }
+                }
+            }
+
+            /* Check if every privilege that is present in either token is also present in the other one */
+            for (FirstTokenIndex = 0; FirstTokenIndex < FirstToken->PrivilegeCount; FirstTokenIndex++)
+            {
+                for (SecondTokenIndex = 0; SecondTokenIndex < SecondToken->PrivilegeCount; SecondTokenIndex++)
+                {
+                    /* Are the privileges from either token equal? */
+                    if (FirstToken->Privileges[FirstTokenIndex].Luid.LowPart !=
+                        SecondToken->Privileges[SecondTokenIndex].Luid.LowPart)
+                    {
+                        DPRINT1("SepCompareTokens(): Privileges are not equal!\n");
+                        goto Quit;
+                    }
+
+                    /* Are the privilege attributes from either token equal? */
+                    if (FirstToken->Privileges[FirstTokenIndex].Attributes !=
+                        SecondToken->Privileges[SecondTokenIndex].Attributes)
+                    {
+                        DPRINT1("SepCompareTokens(): Privilege attributes are not equal!\n");
+                        goto Quit;
+                    }
+
+                    IsEqual = TRUE;
+                    DPRINT1("SepCompareTokens(): Tokens are equal!\n");
+                }
+            }
         }
+    }
+    else
+    {
+        /* If none of the tokens are restricted then just begin checking for every privilege present */
```

Comment:
This code is the same as the one above (that I mark with "****"), so you can group all that together.